codebase2/codebase-sqlite/U/Codebase/Sqlite/LocalizeObject.hs

```diff
@@ -62,70 +68,92 @@ import Unison.Util.Set qualified as Set
 
 -- | Localize a branch object.
 localizeBranch :: DbBranch -> (BranchLocalIds, LocalBranch)
-localizeBranch (Branch terms types patches children) =
+localizeBranch = localizeBranchG
+
+-- | Generalized form of 'localizeBranch'.
+localizeBranchG :: forall t d p c. (Ord t, Ord d, Ord p, Ord c) => Branch' t d p c -> (Branch.BranchLocalIds' t d p c, LocalBranch)
+localizeBranchG (Branch terms types patches children) =
   (runIdentity . runLocalizeBranch) do
     Branch
       <$> Map.bitraverse localizeText (Map.bitraverse localizeReferent localizeBranchMetadata) terms
       <*> Map.bitraverse localizeText (Map.bitraverse localizeReference localizeBranchMetadata) types
       <*> Map.bitraverse localizeText localizePatchReference patches
       <*> Map.bitraverse localizeText localizeBranchReference children
   where
-    localizeBranchMetadata :: (ContainsDefns s, ContainsText s, Monad m) => Branch.DbMetadataSet -> StateT s m Branch.LocalMetadataSet
+    localizeBranchMetadata ::
+      Branch.MetadataSetFormat' t d ->
+      State (LocalizeBranchState t d p c) (Branch.MetadataSetFormat' LocalTextId LocalDefnId)
     localizeBranchMetadata (Branch.Inline refs) =
       Branch.Inline <$> Set.traverse localizeReference refs
 
 -- | Localize a patch object.
 localizePatch :: Patch -> (PatchLocalIds, LocalPatch)
-localizePatch (Patch termEdits typeEdits) =
+localizePatch = localizePatchG
+
+localizePatchG :: forall t h d. (Ord t, Ord h, Ord d) => Patch' t h d -> (PatchLocalIds' t h d, LocalPatch)
+localizePatchG (Patch termEdits typeEdits) =
   (runIdentity . runLocalizePatch) do
     Patch
       <$> Map.bitraverse localizeReferentH (Set.traverse localizeTermEdit) termEdits
       <*> Map.bitraverse localizeReferenceH (Set.traverse localizeTypeEdit) typeEdits
   where
-    localizeTermEdit :: (ContainsText s, ContainsDefns s, Monad m) => TermEdit -> StateT s m LocalTermEdit
+    localizeTermEdit :: (TermEdit' t d) -> State (LocalizePatchState t h d) LocalTermEdit
     localizeTermEdit =
       bitraverse localizeText localizeDefn
 
-    localizeTypeEdit :: (ContainsText s, ContainsDefns s, Monad m) => TypeEdit -> StateT s m LocalTypeEdit
+    localizeTypeEdit :: TypeEdit' t d -> State (LocalizePatchState t h d) LocalTypeEdit
     localizeTypeEdit =
       bitraverse localizeText localizeDefn
 
 --------------------------------------------------------------------------------------------------------------------------------------------
 -- General-purpose localization
 
 -- Contains references to branch objects.
-type ContainsBranches s =
-  HasType (Map (BranchObjectId, CausalHashId) LocalBranchChildId) s
+class Ord c => ContainsBranches c s where
```

Comment:
Need typeclasses here in order to generalize, but these don't escape the module.

Reply:
Ah, ok.